### ISSUE TITLE
Rework stream formation to better detect dependent operations

### ISSUE THREAD

**Issue by benvanik** [google/iree#1882]
Currently all tie_shape ops are being forced into streams: https://github.com/google/iree/blob/6faf80410fb381556f07d906c235b70ce344c770/iree/compiler/Dialect/Flow/Transforms/FormStreams.cpp#L47-L49

This has the side-effect of creating empty streams that perform no real work:
```mlir
    %cmd = hal.command_buffer.create %dev, "OneShot", "Transfer|Dispatch" : !hal.command_buffer
    hal.command_buffer.begin %cmd
    hal.command_buffer.end %cmd
    hal.ex.submit_and_wait %dev, %cmd
```

We should instead be looking for ops that *must* be in streams (StreamableOpInterface::isStreamOnly) and then trying to gather/possibly duplicate ops that are stream-compatible (StreamableOpInterface::isUsableInStream). We can also reorder ops to pull work above/below non-streamable ops instead of just splitting like we are now.

A better algorithm for constructing stream fragments will help #1161.

**Comment by GMNGeoffrey** [comment]
What about a canonicalization (or just a pass if we think this isn't always positive) that drops streams that have no ops that must be streamed (or all ops that "do no work")? Then stream formation doesn't need to worry quite as much about creating these useless streams. Better stream formation also sounds good, but when we can separate concerns that would be nice. I just simplified the stream formation a bunch because it was trying to do a lot all in one pass and ended up being buggy (https://github.com/google/iree/issues/1754).

**Comment by GMNGeoffrey** [comment]
Regarding "possibly duplicate" in https://github.com/google/iree/issues/1601 I think I'm seeing an issue where the stream fragment returns a dynamic shape but it's moved the tie_shape into the fragment so things using it outside can't see it

**Comment by benvanik** [comment]
We need to form the streams without interleaving passes - as moving an op into one stream may prevent another stream from being created.

**Comment by GMNGeoffrey** [comment]
Why not both? A pattern to drop useless streams is easy and frees stream formation from having to check it doesn't create such streams.

**Comment by benvanik** [comment]
Because then we have two places that need to track this logic identically instead of one. Not forming empty streams means the second duplicated bit of code is never hit and not useful.

**Comment by GMNGeoffrey** [comment]
On the subject of tie_shape not always being brought into stream fragments, it looks like we don't have any handling for tie_shape outside of stream fragments atm: https://github.com/google/iree/issues/1894

**Comment by benvanik** [comment]
I'll let you take care of this given your recent changes and #1894 :)

**Comment by rsuderman** [comment]
I believe I encountered this bug for the tensorlist work. My VM operation returns an allocated and filled buffer that is believed to be dynamically shaped (its not but shape propagation doesn't know that).

See the attached input file and the used compiled command
*****
blaze run //integrations/tensorflow/compiler:iree-tf-opt -- --pass-pipeline='canonicalize, func(xla-legalize-control-flow, iree-flow-hlo-to-hlo-preprocessing), iree-flow-flatten-tuples-in-cfg, inline{disable-simplify=false max-iterations=4}, func(canonicalize, cse), iree-flow-legalize-input-types, func(iree-flow-materialize-exported-reflection, iree-shape-expand-function-dynamic-dims, iree-flow-merge-exported-reflection, iree-shape-tie-dynamic, iree-shape-materialize-calculations, iree-shape-hoist-shape-calculations, iree-flow-pre-partitioning-conversion), iree-flow-dispatchability-analysis, func(iree-flow-identify-dispatch-regions, cse, iree-flow-fold-compatible-dispatch-regions, iree-flow-rematerialize-dispatch-constants), iree-flow-outline-dispatch-regions, func(canonicalize, iree-flow-post-partitioning-conversion, iree-flow-hoist-unstreamable-ops, iree-flow-form-streams)' [reproduce.txt](https://github.com/google/iree/files/4646851/reproduce.txt)
*****

Note the [generated IR](https://github.com/google/iree/files/4646877/result.txt) below creates a stream fragment that doesn't depends on the contents of concat.

**Comment by GMNGeoffrey** [comment]
Can you try creating a local crash reproducer that doesn't require TF? https://mlir.llvm.org/docs/PassManagement/#local-reproducer-generation

**Comment by GMNGeoffrey** [comment]
Thanks for the additional example, Rob. I think we should prioritize a stop-the-bleeding fix here so that we don't have people getting random runtime errors. Doing this right will take a bit (especially since a `tie_shape` outside of a stream isn't supported right now). I think we could have a `isStreamEmpty` function that attached to a pattern deleting them and that could also be used in this pass to verify it didn't mess up (or deleted). Probably not good as a canonicalization pattern because it means stream fragments have to know about shape ops (though maybe that's ok?).

**Comment by benvanik** [comment]
Not sure what you mean by isStreamEmpty, but I'm pretty sure that's no good. Can you expand?

**Comment by GMNGeoffrey** [comment]
```c++
static bool isStreamEmpty(ExStreamFragmentOp streamOp) {
  for (auto op : streamOp.getBody().getOps()) {
    if (!isShapeMetadataOp(op)) {
      return false;
    }
  }
  return true;
}
```

This isn't reworking the streams as you were describing, it's just a reusable way of saying "this stream shouldn't exist" which we can use now to avoid hard-to-debug runtime errors and can use later to validate the stream formation output.

**Comment by benvanik** [comment]
nah, that's the wrong order - you want to not create empty streams - creating the stream itself introduces a lot of implicit assumptions that can't be easily unwound. Run that before on the ops that would be put in a stream and just don't create it.

**Comment by benvanik** [comment]
(bandaids are fine, but dirty bandaids are not ;)

**Comment by GMNGeoffrey** [comment]
Yeah fair enough. Bandaids is exactly what I'm looking for since I'd like to quickly get to a point where this isn't causing poor folks like Rob to spend a bunch of time debugging empty-buffer runtime errors. Not creating those types of streams should fix these issues I think (assuming it doesn't break anything else :crossed_fingers:)

Doing this right (as described in the OP) is going to require rewriting quite a bit of stream formation and will make things that are invalid in other ways (#1894).

I think I might philosophically disagree with you on some points though. Since a metadata-only stream is valid IR we should have some kind of error or something. The buffer allocation is assuming the streams aren't empty, which means either that should be verifier-invalid IR or it should emit an error at that point: https://mlir.llvm.org/getting_started/DeveloperGuide/#assertions-and-crashes-in-passes

> If a pass requires additional invariants not guaranteed by the verifier, then it should check them itself and if those invariants are not present, either safely perform no transformation (for pure optimization passes) or emit a diagnostic explaining why a transformation cannot be performed (for lowering passes where the transformation is needed for correctness).

Basically the buffer allocation is a miscompile right now IMO and we should be getting a compiler diagnostic rather than a runtime failure.

**Comment by benvanik** [comment]
The entirety of the stream formation code is throw-away - hence the "ex[perimental]" in the op name. It's fine to make it throw it an error and move on. Keep the bandaids local and easy to throw away with the rest of the work (that's why adding canonicalizations that happen in later passes is bad: it starts to spread the behavior around in ways that get tangled quickly).

**Comment by GMNGeoffrey** [comment]
@benvanik sent a CL your way

**Comment by benvanik** [comment]
(bug hit here was fixed, feature will be done holistically as part of stream IR reworking)
